pkg/controller/daemon/update.go

```diff
@@ -79,6 +79,73 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet, hash st
 	return dsc.syncNodes(ds, oldPodsToDelete, []string{}, hash)
 }
 
+// surgingRollingUpdate creates new daemon set pods to replace old ones making sure that no more
+// than ds.Spec.UpdateStrategy.SurgingRollingUpdate.MaxSurge extra pods are scheduled at any
+// given time.
+func (dsc *DaemonSetsController) surgingRollingUpdate(ds *extensions.DaemonSet, hash string) error {
+	nodeToDaemonPods, err := dsc.getNodesToDaemonPods(ds)
+	if err != nil {
+		return fmt.Errorf("couldn't get node to daemon pod mapping for daemon set %q: %v", ds.Name, err)
+	}
+	maxSurge, numSurge, err := dsc.getSurgeNumbers(ds, nodeToDaemonPods)
+	if err != nil {
+		return fmt.Errorf("Couldn't get surge numbers: %v", err)
+	}
+
+	glog.V(4).Infof("Surging new pods and deleting obsolete old pods")
+	var nodesToSurge []string
+	var oldPodsToDelete []string
+	for node, pods := range nodeToDaemonPods {
+		var newPod, oldPod *v1.Pod
+		for _, pod := range pods {
+			if util.IsPodUpdated(ds.Spec.TemplateGeneration, pod, hash) {
+				if newPod != nil {
+					glog.V(4).Infof("Multiple new pods on node %s: %s, %s", node, newPod.Name, pod.Name)
+				}
+				newPod = pod
+			} else {
+				if oldPod != nil {
+					glog.V(4).Infof("Multiple old pods on node %s: %s, %s", node, oldPod.Name, pod.Name)
+				}
+				oldPod = pod
+			}
+		}
+
+		if newPod == nil && numSurge < maxSurge {
```

Comment:
Please add some comments to these two conditions. The first is pretty simple, but the second could use some explanation.